### PR TITLE
Print feature Proof-of-Concept (episode 2)

### PR DESCRIPTION
**Pull request by herbelin**
This is a continuation of #85. See there for the discussion.

### CODE REVIEW

File: toplevel/toplevel.mllib [hunk 1]
```diff
@@ -1,3 +1,4 @@
+Extravernac
```
Comment (gares) [inline toplevel/toplevel.mllib]: What is that?

**Comment by gares** [comment]
IIRC the conclusion of the discussion was to merge the hooks and let people use them in the plugins.
Then the original PR had no hooks, so nothing was merged.

This PR adds the hooks and the plugin. IMHO the plugin should be provided as an OPAM package, while the hooks merged.

**Comment by maximedenes** [comment]
> IIRC the conclusion of the discussion was to merge the hooks and let people use them in the plugins.

That's also my recollection of the discussion.

**Comment by herbelin** [comment]
I'm convinced that having a printing side-effect, but also say a read side-effect, etc. is of general interest in everyday use of Coq, starting with debugging. I don't see the point of having to install an external plugin.

**Comment by Zimmi48** [comment]
> IMHO the plugin should be provided as an OPAM package, while the hooks merged.

What is the argument for having it in an external plugin?

@herbelin can't you just commit be04afd to trunk directly? It doesn't relate to this PR as far AFAICS and can't possibly break anything.

**Comment by gares** [comment]
I'm just recalling what was discussed on the PR that was cited and the conclusion that was reached.
And that this PR is not implementing what was concluded there (it does more).

Now you (two) are questioning the conclusion of that discussion. That is legitimate, but I guess we should then discuss what are the new arguments in favor of putting the code in Coq, or how/why the conclusion that was reached is not valid anymore. I mean, the context may have changed, new arguments may have appeared...

**Comment by Zimmi48** [comment]
@gares OK, sorry. I thought the problem was plugin vs not plugin rather than in Coq vs external.

**Comment by maximedenes** [comment]
Like @gares said, this PR does not implement what we agreed on when discussing it before the 8.6 freeze.  I'm not aware of any new discussion having taken place. If you want it merged, please separate the infrastructure (on Coq's side) from the plugin. I'll merge the former after review. As they are, the commits don't separate the plugin from the rest.

**Comment by herbelin** [comment]
I did not follow precisely the discussion about #85 at the 8.6 freeze. I asked [later on](https://github.com/coq/coq/pull/85#issuecomment-248046492) for details to which I got an answer. I asked [here](https://github.com/coq/coq/pull/85#issuecomment-248056823) and [there](https://github.com/coq/coq/pull/85#issuecomment-284989608) about how to proceed and noone told I was misunderstanding anything.

Can someone give more details about the discussion? I'm asking because I really think this feature has a great potential, starting for debugging purpose. There was a discussion about implementing it more than 10 years ago already. In some sense, it is also superior in terms of what it allows to what the option `Set Debug Cbv` provides and there was no objection to integrating this later option.

So, any help would be useful! Thanks in advance.

**Comment by maximedenes** [comment]
> I did not follow precisely the discussion about #85 at the 8.6 freeze.

Well, we did have some exchanges at that time, though.

> Can someone give more details about the discussion?

The only remaining misunderstanding between the conclusion of that discussion and what this PR implements is that the the plugin should be shipped externally (an OPAM package would be good). The motivation was to make it possible to get feedback from users, refine the design and the implementation, and integrate once it is mature.

**Comment by herbelin** [comment]
Ok, that's not a bad idea, let's do like this. Moreover, this will be a good incitative to resuscitate this idea to embed opam into Coq so that "Require Import PrintingEffect" works on the nose, making the life of using packages easier. (Please tell if you have some idea on how to do that, so that we can join forces.)

I shall prepare an opam package. In the meantime, I'm removing the `fix_needed` flag for listing in the 8.7 inclusion list.

**Comment by herbelin** [comment]
The plugin part has been moved to an independent [package](https://github.com/herbelin/reduction-effects) which I shall eventually submit as an opam package.

Currently, this opam package only provides a generic `print` of type `forall A, A -> unit` which prints its second argument and a generic `print_id` of type `forall A, A -> A` which also prints its second argument.

I would find cool to provide things like e.g.:
```coq
Definition f x y :=
  let _ := print_string "The result of " in
  let _ := print_nat x in
  let _ := print_string " and " in
  let _ := print_nat y in
  let _ := print_string " is " in
  let _ := print_nat (x+y) in
  let _ := print_string "\n" in
  tt.
Eval cbv in f 3 4.
```
However, two problems arise: 
- the generic printer writes quotes around strings and this is not what is expected here
- the generic printer also always outputs a newline.

Not outputting a newline raises the question of what it would mean in the context of a UI model. What is the best approach to output incremental messages, taking time into account, successively seeing, e.g.: `The result of 3+4 is`, some time passing, `The result of 3+4 is 7`?

I wondered whether providing a type of output for `print` would make sense, say a type:
```coq
Inductive printable := String : string -> printable | Generic : forall A, A -> printable.
Inductive print_command := WithNewLine : list printable -> print_command | WithoutNewLine : list printable -> print_command.
```
and having some `print` effect taking a `print_command` as argument. Any opinion, or better idea?

**Comment by Zimmi48** [comment]
Hi @herbelin, just a small comment in passing on your new plugin which I'm putting here (instead of telling you in person) so that everyone sees it (since this is general advice):
Please do not make the same mistake that was made when licensing Coq of licensing it under LGPL 2.1 only. Instead you can use the formulation [proposed in the license text](https://github.com/herbelin/reduction-effects/blob/master/LICENSE#L476-L479) "version 2.1 of the License, or (at your option) any later version" both in the file headers and in a README in the repo.
Since this plugin is a new and independent work, there is nothing preventing you to do so.

**Pull request merged**